models/build.php

```diff
@@ -513,33 +514,63 @@ public function Save()
             $this->Type = pdo_real_escape_string($this->Type);
             $this->Generator = pdo_real_escape_string($this->Generator);
 
-            $parentId = 0;
+            $this->ParentId = 0;
             $justCreatedParent = false;
             if ($this->SubProjectName) {
-                $parentId = $this->GetParentBuildId();
-                if ($parentId == 0) {
+                $this->ParentId = $this->GetParentBuildId();
+                if ($this->ParentId == 0) {
                     // This is the first subproject to submit for a new build.
                     // Create a new parent build for it.
-                    $parentId = $this->CreateParentBuild($nbuilderrors, $nbuildwarnings);
-                    $justCreatedParent = true;
+                    $justCreatedParent = $this->CreateParentBuild($nbuilderrors, $nbuildwarnings);
                 }
             }
-            $this->ParentId = $parentId;
+            $this->Uuid = Build::GenerateUuid($this->Stamp, $this->Name,
+                    $this->SiteId, $this->ProjectId, $this->SubProjectName);
 
             $query =
                 "INSERT INTO build
                 (".$id."siteid, projectid, stamp, name, type, generator,
                  starttime, endtime, submittime, command, log, builderrors,
-                 buildwarnings, parentid)
+                 buildwarnings, parentid, uuid)
                 VALUES
                 (".$idvalue."'$this->SiteId', '$this->ProjectId',
                  '$this->Stamp', '$this->Name', '$this->Type',
                  '$this->Generator', '$this->StartTime', '$this->EndTime',
                  '$this->SubmitTime', '$this->Command', '$this->Log',
-                 $nbuilderrors, $nbuildwarnings, $this->ParentId)";
+                 $nbuilderrors, $nbuildwarnings, $this->ParentId,
+                 '$this->Uuid')";
 
             if (!pdo_query($query)) {
-                add_last_sql_error("Build Insert", $this->ProjectId, $this->Id);
+                $error = pdo_error();
+                // This error might be due to a unique constraint violation
+                // for this UUID.  Query for such a previously existing build.
```

Comment:
Expecting unique key violations (as we do here) will not work well with @cpatrick 's exponential backoff feature (#90).  Any ideas on how to get these two to play nicely together?